targettests/Remote-Sim/args_synthesis.cpp

```diff
@@ -0,0 +1,559 @@
+    rx(1.0 * k[i], q);
+}
+
+// List of all the test functors to run
+static std::vector<std::function<void()>> ALL_TEST_FUNCTORS;
+
+// Main macro to define a test with different signatures.
+// - Dispatch each argument to the test function 'f'.
+// - Add an execution functor to ALL_TEST_FUNCTORS.
+// For example:
+// DEFINE_TEST_KERNEL(test, (int, a, INT_VAL)(std::vector<float>, b,
+// VEC_FLOAT_VAL)(bool, c, BOOL_VAL)(std::vector<int>,
+// d,VEC_INT_VAL)(std::vector<double>, e, VEC_DOUBLE_VAL));
+
+// will be expanded to (define a kernel with a signature and add a test functor
+// to the list)
+
+// void test(int a, std::vector<float> b, bool c, std::vector<int> d,
+//                    std::vector<double> e) __qpu__ {
+//   cudaq::qubit q;
+//   f(q, a);
+//   f(q, b);
+//   f(q, c);
+//   f(q, d);
+//   f(q, e);
+//   mz(q);
+// }
+// const bool addedtest = []() {
+//   ALL_TEST_FUNCTORS.emplace_back([]() {
+//     auto counts = cudaq::sample(test, INT_VAL, VEC_FLOAT_VAL, BOOL_VAL,
+//     VEC_INT_VAL, VEC_DOUBLE_VAL); counts.dump();
+//   });
+//   return true;
+// }();
+
+#define DEFINE_TEST_KERNEL(func_, ...)                                         \
+  void func_(END(PARAMS_LOOP_0 __VA_ARGS__)) __qpu__ {                         \
+    cudaq::qubit q;                                                            \
+    END(VAR_LOOP_0 __VA_ARGS__)                                                \
+    mz(q);                                                                     \
+  }                                                                            \
+  const bool added##func_ = []() {                                             \
+    ALL_TEST_FUNCTORS.emplace_back([]() {                                      \
+      std::cout << "Testing: " << #func_ << "\n";                              \
+      auto counts = cudaq::sample(func_ END(INVOKE_LOOP_0 __VA_ARGS__));       \
+      counts.dump();                                                           \
+    });                                                                        \
+    return true;                                                               \
+  }();
+
+// Define some dummy argument values to use for testing.
+static const bool BOOL_VAL = true;
+static const char CHAR_VAL = 'a';
+static const int INT_VAL = 1;
+static const std::size_t SIZE_T_VAL = 1;
+static const int64_t INT64_VAL = 1;
+static const double DOUBLE_VAL = 1.0;
+static const float FLOAT_VAL = 1.0;
+static const std::vector<bool> VEC_BOOL_VAL = {true, false};
+static const std::vector<char> VEC_CHAR_VAL = {'a', 'b'};
+static const std::vector<int> VEC_INT_VAL = {1, 2};
+static const std::vector<std::size_t> VEC_SIZE_T_VAL = {1, 2};
+static const std::vector<double> VEC_DOUBLE_VAL = {1.0, 2.0};
+static const std::vector<float> VEC_FLOAT_VAL = {1.0, 2.0};
+
+// Define tests
+// Note: these are single-qubit tests, so will run very fast.
+DEFINE_TEST_KERNEL(test1, (bool, a, BOOL_VAL));
```

Comment:
When one of these tests fails, will we be able to easily tell which one is failing? My quick test of uncommenting out test that we expect to fail (e.g. `bool`) left me with a stack trace that wasn't immediately obvious to me which was failing.

Reply:
For those runtime failures (`vector<char>` in the `QuakeSynthesizer`), we do see a log about the name of the test. 

In the case of compiler errors (`vector<bool>`), unfortunately, I don't have a good solution for the logging. 
BTW, to see the actual pass error of `vector<bool>`, we also need to uncomment the overload for `vector<bool>` (line 87-91)

Reply:
Can we use templates here instead of the macros? (The macros are impressive though.)

Reply:
I guess we can. In fact, I've tried to used templates but gave up after an hour or so :) 

The primary reason of using macros is that I can debug the generated code more quickly. Plus, I think the generated code would also mimic the end-users' usage that we want to test (writing kernels with concrete types, not having an extra template instantiation step to be handled by the bridge, etc.)

Reply:
This sounds interesting. We handle templates in other places, so I'm a bit curious about the problems you saw.

Reply:
> The primary reason of using macros is that I can debug the generated code more quickly. Plus, I think the generated code would also mimic the end-users' usage that we want to test (writing kernels with concrete types, not having an extra template instantiation step to be handled by the bridge, etc.)

That's true. And being able to compile with `-E` has great benefits here, in my opinion.

> In the case of compiler errors (vector<bool>), unfortunately, I don't have a good solution for the logging.
BTW, to see the actual pass error of vector<bool>, we also need to uncomment the overload for vector<bool> (line 87-91)

Re: identifying specific test cases when failures occur, would it be possible to split the tests into groups to reduce the amount of commenting and uncommenting that would be required to identify the failure?

Reply:
Thanks @bmhowe23 for the suggestion. I followed a similar approach, but split the tests into headers instead (https://github.com/NVIDIA/cuda-quantum/pull/1385/commits/fce39625860e3e6ab73d2dda9fe1eef42ed43a19) 

Please let me know if it addresses the failure analysis need. I also added source location info of each test in the runtime logging.

@schweitzpgi I don't have a specific reproducer. It might well be a convoluted failure early on where we had these issues with small types, which have since been fixed.

Reply:
I see you later went with the `#define`'s...that looks good to me, so I'm marking this conversation as resolved.